### PR TITLE
fix macOS spawn default

### PR DESCRIPTION
**Pull request by mattreex**
https://jira.opensciencedatacloud.org/browse/DEV-809
Python 3.8 set the default multiprocessing start mode to spawn for macOS. Fork is default for ubuntu. Assert fork mode is used for macOS as well.

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -1,6 +1,6 @@
 #!groovy
 
-library identifier: "jenkins-lib@develop"
+library identifier: "jenkins-lib@master"
```
Comment (philtom-ctds) [inline Jenkinsfile:3]: What's the purpose of this change?

File: indexd_test_utils/__init__.py [hunk 2]
```diff
@@ -149,8 +151,13 @@ def indexd_server():
     hostname = 'localhost'
     port = 8001
     debug = False
+    proc_handler = mp.Process
+    if six.PY3:
+        # Note: explicitly specifying fork, as spawn is the default in
+        # py3.8+ on macos.
+        proc_handler = mp.get_context("fork").Process
```
Comment (philtom-ctds) [inline indexd_test_utils/__init__.py:158]: It looks like this issue runs deeper.

The official python docs state that "fork" is considered unsafe on macOS:
https://docs.python.org/3.8/library/multiprocessing.html#contexts-and-start-methods

The change to make "spawn" the default for macOS is documented and discussed in this ticket:
https://bugs.python.org/issue33725

Some issues related to "spawn" on macOS are documented here:
https://bugs.python.org/issue40106

Could you investigate this further to see what our options are?

Btw, if we were to stick with using fork, we should also be explicit and check that the os is macOS.   https://docs.python.org/3/library/sys.html#sys.platform

File: Jenkinsfile (hunk 1 above)
Comment (ProfOak) [reply Jenkinsfile:3]: All of our other services use the master branch of jenkins-lib. The develop branch is behind master by a bit.

File: indexd_test_utils/__init__.py (hunk 2 above)
Comment (mattreex) [reply indexd_test_utils/__init__.py:158]: Other than the current solution, we could mock indexd. I am not sure how much of a concern this is given that it is only tests. This issue in mac isn't new https://bugs.python.org/msg344590, but macOS has become more aggressive at detecting this and crashing early. I've encountered this solution in other popular python libraries, including pytorch https://github.com/pytorch/pytorch/pull/36687 . 

Moreover, Werkzeug only provides support for fork https://github.com/pallets/werkzeug/blob/d781b06c3fbac8eaeb3bfb61bea005a3a5288b91/src/werkzeug/serving.py#L763 . After some thought, it is not clear there is a good way to get around this that doesn't require a large effort.

Comment (kulgan) [reply indexd_test_utils/__init__.py:158]: Since this is only applicable to tests, we can table this for now and make a separate ticket to rewrite this section without using multiprocessing or something like that. I also think this might affect other projects by the time we update them to py38

**Review by kulgan — APPROVED** [review]

File: indexd_test_utils/__init__.py [hunk 3]
```diff
@@ -149,18 +149,12 @@ def indexd_server():
...
-
-    indexd = Process(
-        target=app.run,
-        args=(hostname, port),
-        kwargs={'debug': debug},
-    )
-    indexd.start()
+    t = threading.Thread(target=app.run, kwargs={'host': hostname, 'port': port, 'debug': debug})
+    t.setDaemon(True)
+    t.start()
```
Comment (kulgan) [inline indexd_test_utils/__init__.py:154]: I did a couple of tests mainly to check if the server stopped successfully after running any test, and it behaved correctly thanks to the `setDaemon`. But, it now means that this solution will only work for `session` scoped fixtures since there is no explicitly defined way to stop the thread. Luckily this is a `session` scoped fixture which I don't see changing any time soon, so this is good enough for me

**Review by kulgan — APPROVED** [review]

File: indexd_test_utils/__init__.py (hunk 3 above)
Comment (ProfOak) [reply indexd_test_utils/__init__.py:154]: Should this information be included as a comment in the code?

**Review by ProfOak — APPROVED** [review]
Seems ok to me. I wondered if there was a speed difference in test execution with threading vs multiprocessing. I checked the test runs between this branch and the current develop, and there appears to be no difference.

File: indexd_test_utils/__init__.py (hunk 3 above)
Comment (mattreex) [reply indexd_test_utils/__init__.py:154]: I think it's explicit in the code.
